### PR TITLE
Add diagnostic for pluscal labels inserted by translation

### PR DESCRIPTION
**Pull request by hwayne**
The PlusCal translator has a `-reportLabels` flag. When set, it will automatically add missing labels to translations and log that in the output:

```
pcal.trans Version 1.11 of 31 December 2020
The following label was added:
  Lbl_1 at line 16, column 5
Parsing completed.
Translation completed.
```

Since it's logged, we might as well tell the user about it! This PR adds an information diagnostic. Demo:

image

image

Questions: 

- Should it be a `warning` instead of `information`?
- In the long run should the pluscal parser use some kind of state machine?

### CODE REVIEW

File: src/parsers/pluscal.ts [hunk 1]
```diff
@@ -91,11 +94,49 @@ export class TranspilerStdoutParser extends ProcessOutputHandler<DCollection> {
         return true;
     }
 
+
```
Comment (lemmy) [inline src/parsers/pluscal.ts:97]: Please minimize unnecessary whitespace, extra newlines, or similar changes to ensure the diff remains small.

File: src/parsers/pluscal.ts [hunk 2]
```diff
@@ -23,6 +23,7 @@ export class TranspilerStdoutParser extends ProcessOutputHandler<DCollection> {
     private readonly filePath: string;
     private errMessage: string | null = null;
     private nextLineIsError = false;
+    private nowParsingAddedLabels = false; //should probably be a state machine thing
```
Comment (lemmy) [inline src/parsers/pluscal.ts:26]: Is the comment a TODO? If so, could you provide more detail and make it clearer?

File: src/parsers/pluscal.ts [hunk 3]
```diff
@@ -91,11 +94,49 @@ export class TranspilerStdoutParser extends ProcessOutputHandler<DCollection> {
...
+     *       The following labels were added: // 1
+     *           Lbl_1 at line 16, column 5
+     *           Lbl_2 at line 23, column 9
+     * 
+     * 
+     * So we can start looking for labels as soon as we see (1)
+     * and stop as soon as we stop seeing label strings.
+     */
+    private tryParseAddedLabels(line: string) {
+        if (line.startsWith("The following label")) { // could be `was added` or `were added`
```
Comment (lemmy) [inline src/parsers/pluscal.ts:117]: Why not use a regex here as well, considering you're already using a regex matcher further down?

Moreover, I'd suggest to add a link to the PCal method that emits these strings: https://github.com/tlaplus/tlaplus/blob/21f924cc0d77501ae2de104dda4f56997a29bc5b/tlatools/org.lamport.tlatools/src/pcal/ParseAlgorithm.java#L668-L682

File: src/parsers/pluscal.ts [hunk 4]
```diff
@@ -91,11 +94,49 @@ export class TranspilerStdoutParser extends ProcessOutputHandler<DCollection> {
...
+        }
+
+        const matcher = /^\s\s([A-Za-z0-9_]+) at line (\d+), column (\d+)$/g.exec(line)
+        if (!matcher) { // done parsing 
+            this.nowParsingAddedLabels = false;
+            return false;
+        }
+
+        const labelname = matcher[1];
+        const message = `Missing label, translator inserted \`${labelname}\` here`;
```
Comment (lemmy) [inline src/parsers/pluscal.ts:132]: https://github.com/tlaplus/tlaplus/blob/21f924cc0d77501ae2de104dda4f56997a29bc5b/tlatools/org.lamport.tlatools/src/pcal/ParseAlgorithm.java#L654-L666

File: src/parsers/pluscal.ts [hunk 5]
```diff
@@ -91,11 +94,49 @@ export class TranspilerStdoutParser extends ProcessOutputHandler<DCollection> {
...
+    private tryParseAddedLabels(line: string) {
+        if (line.startsWith("The following label")) { // could be `was added` or `were added`
+            this.nowParsingAddedLabels = true;
+            return true;
+        };
+        if (!this.nowParsingAddedLabels) {
+            return false
+        }
+
+        const matcher = /^\s\s([A-Za-z0-9_]+) at line (\d+), column (\d+)$/g.exec(line)
```
Comment (lemmy) [inline src/parsers/pluscal.ts:125]: Is this robust against any accepted "label root" that a user may pass with `-labelRoot`?

**Review by lemmy — APPROVED** [review]
See comments

File: src/parsers/pluscal.ts (hunk 3 above)
Comment (hwayne) [reply src/parsers/pluscal.ts:117]: We need to use a regex to capture the name of the label below. Here we don't need to capture anything, so it makes sense to use something weaker. It also follows the existing convention on line 69 https://github.com/tlaplus/vscode-tlaplus/pull/350/files#diff-5395c8cd3da8e88aabad5e743d78366c8848a3a4ffa280ed77b08c02a0c6e7a4R69

File: src/parsers/pluscal.ts (hunk 4 above)
Comment (lemmy) [reply src/parsers/pluscal.ts:133]: No, lets just add a code comment as a pointer to where Pcal emits this string.

File: src/parsers/pluscal.ts (hunk 3 above)
Comment (lemmy) [reply src/parsers/pluscal.ts]: It is possible that Pcal will someday emit something like "The following label was removed".  In order to make the code more robust, how about: 

```typescript
const validPrefixes = ["The following labels were added:", "The following label was added:"];

if (validPrefixes.some(prefix => line.startsWith(prefix))) {
    ...
}
```

Comment (hwayne) [reply src/parsers/pluscal.ts]: Done!

File: src/parsers/pluscal.ts (hunk 5 above)
Comment (lemmy) [reply src/parsers/pluscal.ts]: Can you please open a Pcal issue?

Comment (hwayne) [reply src/parsers/pluscal.ts]: Took some time but that's now done!

Comment (hwayne) [reply src/parsers/pluscal.ts]: https://github.com/tlaplus/tlaplus/issues/1092
